public/app/features/datasources/components/picker/DataSourceModal.test.tsx

```diff
@@ -0,0 +1,110 @@
+import { render, screen } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import React from 'react';
+
+import { DataSourceInstanceSettings, DataSourcePluginMeta, PluginMetaInfo, PluginType } from '@grafana/data';
+import { GetDataSourceListFilters } from '@grafana/runtime';
+
+import { DataSourceModal } from './DataSourceModal';
+
+const pluginMetaInfo: PluginMetaInfo = {
+  author: { name: '' },
+  description: '',
+  screenshots: [],
+  version: '',
+  updated: '',
+  links: [],
+  logos: { small: '', large: '' },
+};
+
+function createPluginMeta(name: string, builtIn: boolean): DataSourcePluginMeta {
+  return { name, id: name, type: PluginType.datasource, baseUrl: '', info: pluginMetaInfo, module: '' };
+}
+
+function createDS(name: string, id: number, builtIn: boolean): DataSourceInstanceSettings {
+  return {
+    name: name,
+    uid: name + 'uid',
+    meta: createPluginMeta(name, false),
+    id,
+    access: 'direct',
+    jsonData: {},
+    type: '',
+    readOnly: true,
+  };
+}
+
+const mockDS = createDS('mockDS', 1, false);
+const xMockDS = createDS('xMockDS', 2, false);
+const builtInMockDS = createDS('builtInMockDS', 3, true);
+
+const mockDSList = [mockDS, xMockDS, builtInMockDS];
+
+const setup = (onChange = () => {}, onDismiss = () => {}) => {
+  const props = { onChange, onDismiss, current: mockDS.name };
+  window.HTMLElement.prototype.scrollIntoView = function () {};
+  return render(<DataSourceModal {...props}></DataSourceModal>);
+};
+
+jest.mock('@grafana/runtime', () => {
+  const actual = jest.requireActual('@grafana/runtime');
+  return {
+    ...actual,
+    getTemplateSrv: () => {
+      return {
+        getVariables: () => [],
+      };
+    },
+  };
+});
+
+jest.mock('@grafana/runtime/src/services/dataSourceSrv', () => {
+  return {
+    getDataSourceSrv: () => ({
+      getList: (filters: GetDataSourceListFilters) => {
+        if (filters.filter) {
+          return mockDSList.filter(filters.filter);
+        }
+        return mockDSList;
+      },
+      getInstanceSettings: () => mockDS,
+    }),
+  };
+});
+
+describe('DataSourceDropdown', () => {
+  it('should render', () => {
+    expect(() => setup()).not.toThrow();
+  });
+
+  describe('interactions', () => {
```

Comment:
There are interactions not contemplated in this PR:
1. Click in "Configure a new DS" works as expected
2. It displays the 3 built-in DS on the right column, but not in the left one
3. When searching and not finding results, it shows the empty state with the CTA to add a new DS

There is some configuration important here we're not testing:
1. When the FF is disabled, we don't display the CSV uploader. When is enabled, the opposite.
2. Should we test the dropping somehow? I think it is an important feature and it has a specific onChange API that is worth documenting through tests so anyone can understand why the second parameter is there.

Reply:
Tests for this added in: https://github.com/grafana/grafana/pull/68236/commits/8449bebf2badec0fcedae596f3d71f93221772d9

I opted not to test the `Configure a new DS` since it's just a link. Testing that it's there should be enough.